### ISSUE TITLE
F2 (full screen) Unconfiguration display seetings in KDE with 2 screens

### ISSUE THREAD

**Issue by lupoDharkael** [ColinPitrat/caprice32#84]
Every time I press F2 in the emulator I have to reconfigure the screen resolution of my main screen because it gets a wrong resolution and position relative to the other screen. The problem is reproducible and persistent (I have to restore the original configuration manually) in my system.

**Comment by ColinPitrat** [comment]
Do you experience the same issue with other games using SDL (maybe only SDL 1.2 ?) in fullscreen mode ?

**Comment by lupoDharkael** [comment]
I haven't played games since I leaved Windows half a year ago, now I use only free software. I'd like to help to identify the problem . I'll test this in another computer with other desktop environment tomorrow. Could you provide the name of other program to test the possible fullscreen issue with SDL? I would like to check it (if you don't know any I could try to find someone)

**Comment by sebhz** [comment]
Hi,
If I get it right you have a multihead configuration with two monitors ("main" and "aux" screen), have caprice32 run on the "aux" screen, and when you send caprice32 to fullscreen, with F2, the resolution of your "main" screen gets corrupted/reset to default/whatever - and stays corrupted when you leave Caprice32. Correct ? In this case Caprice32 is correctly displayed fullscreen on your aux monitor (or so I assume).
I never tried this configuration but can give it a go (unfortunately not before next week) and try and see what the problem is. SDL is not famous for its multihead support, though :-)

**Comment by ColinPitrat** [comment]
Here is a long list of free games using SDL (I'm not sure which ones have a fullscreen mode though):
https://libregamewiki.org/SDL_games

I ran caprice on a dual screen setup in the past and had no issue (except maybe in case of crash). I'll retry it tomorrow. Are you using Wayland ? Which distribution, video card, driver, etc ... ?

**Comment by lupoDharkael** [comment]
I use Fedora 25 (X server) with KDE v5.8.6. I'm running the program with the Mesa driver 13.0.4
After upgrading to KDE 5.9.4 the problem keeps appearing.
I'll be more descriptive about what happens. When I press F2 both screens blink, after that the programs appears in full screen in both screens with the resolution of my second and which has lower resolution than the main one. My main screen adopts the resolution of the second one and they overlap (their relative position)  that causes seeing the emulator in both screens.
After testing it in a game called Chromium BSU it does the same so it seems to be a problem with SDL.

**Comment by ColinPitrat** [comment]
Do you have another window manager to test with ? I remember KDE 4 was
handling screen settings in a strange way (bypassing X) and this could be
the issue here: SDL changes the screen mode and then at exit, X restore the
default one without KDE noticing. I'd suspect a restart of KDE fixes the
display ?

Regards,
Colin

**Comment by lupoDharkael** [comment]
I'm back (sorry for the delay). I've tested it on XFCE in the same machine (I don't have another computer with 2 hdmi or vga ports) and it does the same as in KDE. A restart doesn't fix the problem because it changes my system configuration, I have to reconfigure the primary screen resolution and relative position of the screens manually.

**Comment by sebhz** [comment]
Which video filter did you used ? If this is not already the case, could you try openGL scaling before going full screen. I did see some funny effects when using other filters and going full screen. You can switch to openGL scaling through the UI or by changing the video filter in cap32.cfg.

**Comment by jfebrer** [comment]
I think the problem is that caprice changes the screen resolution when go to full screen. Another problem in full screen is that doesn't keep the aspect ratio and the emulation get a bit distorted on a 16:9 screen. I have tried Caprice on Retroarch and it doesn't has these problems, don't change the screen resolution when go to full screen and it keep the aspect ratio.

**Comment by ColinPitrat** [comment]
Is the display very small then when going full screen ?

**Comment by jfebrer** [comment]
Yes, the Caprice32 window is small, but I tried with different settings, with double size, the window is bigger but when goes to fullscreen it changes the screen resolution and doesn't maintain the aspect ratio.
Also it seems to ignore the values scr_width and scr_height on ~/.cap32.cfg, I tried to increase them for example to 1280 and 1024 but Caprice always start with the same window size.

Currenty I use these parameters on the video section of my .cap32.cfg:

[video]
scr_bpp=32
scr_fps=1
scr_width=1280
scr_height=1024
scr_intensity=10
scr_led=1
scr_oglfilter=1
scr_oglscanlines=30
scr_remanency=0
scr_style=1
scr_tube=0
scr_window=1

Thank you for your work on this project!

**Comment by ColinPitrat** [comment]
I meant on retroarch, when you go fullscreen, is the display small or is it really fullscreen. Did you try opengl video output with Caprice32 ? I should maybe make it the default nowaday ...

**Comment by jfebrer** [comment]
Yes on retroarch on fullscreen the display is really fullscreen keeping the 4:3 aspect ratio.
I also tried with the opengl video output on Caprice32 but with the same results as before.

I attach you two screenshots, the first one is retroarch on fullscreen, and the other with the two emulators.

![retroarch-fullscreen](https://user-images.githubusercontent.com/3382653/33406650-33bad2bc-d56e-11e7-9e69-23a2d9f3fe33.png)
![screenshot_20171130_010748](https://user-images.githubusercontent.com/3382653/33406651-33f283ce-d56e-11e7-9337-ede3a6c3e773.png)

**Comment by kamelie1706** [comment]
Hi,

I am running multiscreen and embedded full screen option is not working correctly for me.
For some stange reason caprice32 does not enable KDE to manage the fullscreen/resize of the window.

Not sure if related to this bug or something else ...

**Comment by ColinPitrat** [comment]
Can you provide more details? Screenshots or a screencast of what happens?

**Comment by kamelie1706** [comment]
Hi,

This is maybe more explicit
https://youtu.be/K2zYOQGVPVk

Thx

**Comment by ColinPitrat** [comment]
The inability to resize the window is expected.
The default video plugins wouldn't support resizing.
We could allow resizing for some video plugins but this would require to explicitly set it with SDL_RESIZABLE and handle SDL_VIDEORESIZE events to recall SDL_SetVideoMode.

The fullscreen behavior is weird. Is it trying to center the display across both screen?
My experience with a dual screen setup was that it was setting it fullscreen on the primary display only.
Can you try it with different video plugins? In particular, OpenGL.
You can set it in cap32.cfg with:

```
scr_style=1
```

You can also try setting different resolution for the fullscreen mode:

```
scr_width=1280
scr_height=1024
```

**Comment by kamelie1706** [comment]
Thx I will try that.

Yes if I understand correctly X-windows do not manage "fullscreen by screen" and see both screen as one and only one screen. This is by the way amazing it worked out of the box with 2 video cards!

This "concept" is managed by the window manager, in my case KDE/KWin.

**Comment by kamelie1706** [comment]
Strange

```
scr_width=1280
scr_height=1024
```
Does nothing ... It still launch with default resolution.
I was thinking it was not reading my local conf file ... but no I could switch to monochrome, so reading the customized file.

**Comment by ColinPitrat** [comment]
Sorry, I should have explained that the width and height only impact fullscreen mode. They are the ones used when creating the full screen surface.

For the video plugin, I'm talking about the one in caprice. You can configure it from the GUI or using scr_style in the config file. 11 is OpenGL.

**Comment by ColinPitrat** [comment]
By the way, do you use xorg or Wayland?

**Comment by kamelie1706** [comment]
Great thx. That starts to look better. At least I do not get the "in between" effect.
I am using xorg.

My setup is a bit hybrid
Nvidia 1660 => PCI-E 1 => DVI-D => LCD / Normally main screen
AMD 7450 => PCI-E 2 => VGA => CRT / Secondary screen mostly for emulation

If I disable the LCD screen, the full screen now works correctly on the CRT.
Now if both are active (no matter how I set the primary screen) the full screen appears correctly but only on the LCD or both if I replicate the screens but I cannot get it on the CRT only.

I assume the program picks the first graphic card / first entry it found (in my case the NVIDIA card) to output the fullscreen. 

I do not think it is a problem of Caprice32 anymore as same behaviour with the Hatari emulator when using the embedded fullscreen option. 
The only difference is with Hatari I can simply set resolution in window mode and remove window frame/bar.

**Comment by kamelie1706** [comment]
For reference I have another track to explore related to SDL env variables such as 
export **SDL_VIDEO_FULLSCREEN_DISPLAY**=0

**Comment by kamelie1706** [comment]
Bingo!
Setting this correctly targets the right screen correctly for caprice32
export SDL_VIDEO_FULLSCREEN_DISPLAY=1
